test/cql-pytest/conftest.py

```diff
@@ -93,9 +93,17 @@ def this_dc(cql):
 # and automatically deleted at the end. We use scope="session" so that all
 # tests will reuse the same keyspace.
 @pytest.fixture(scope="session")
-def test_keyspace(cql, this_dc):
+def test_keyspace(request, cql, this_dc):
     name = unique_name()
-    cql.execute("CREATE KEYSPACE " + name + " WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', '" + this_dc + "' : 1 }")
+    stmt = "CREATE KEYSPACE " + name + " WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', '" + this_dc + "' : 1 }"
+    if hasattr(request, "param"):
+        if request.param == "vnodes":
+            stmt += " AND TABLETS = {'enabled': false}"
+        elif request.param == "tablets":
+            stmt += " AND TABLETS = {'enabled': true}"
+        else:
+            pytest.fail(f"test_keyspace(): invalid request parameter: {request.param}")
+    cql.execute(stmt)
```

Comment:
Nice. I can't say, though, how this actually works... Since the test_keyspace fixture is session-scoped, it's supposed to built only once in the entire pytest session. How does it know it needs to create two copies - one for vnodes and for tablets and keep them both?

In fact https://docs.pytest.org/en/6.2.x/fixture.html suggests that it can't:

**Pytest only caches one instance of a fixture at a time, which means that when using a parametrized fixture, pytest may invoke a fixture more than once in the given scope.**

I have an idea: you can do something like I did already in several places in the past: Create two session-scoped fixtures, test_keyspace_vnodes and test_keyspace_tablets, and then have a function-scoped fixture test_keyspace which returns one of these two fixtures as appropriate.

Reply:
I think pytest will keep around one instance for each param and one for the no-param case.
I did a quick experiment and each test received the instance it requested.

I considered your suggestion when writing this, but it lead to uglier code. Using the built-in tools like `pytest.paramtrize()` is so much more elegant.

Here is a quick experiment to prove it:
```python
import pytest

param1 = "param1"
param2 = "param2"
no_param = "no_param"


@pytest.fixture(scope="session")
def parametrized_fixture(request):
    if hasattr(request, "param"):
        if request.param == 1:
            return param1
        else:
            return param2
    return no_param


@pytest.mark.parametrize("parametrized_fixture", [1, 2], indirect=True)
def test_with_param_1(parametrized_fixture):
    assert parametrized_fixture == param1 or parametrized_fixture == param2
    assert parametrized_fixture is param1 or parametrized_fixture is param2


@pytest.mark.parametrize("parametrized_fixture", [1, 2], indirect=True)
def test_with_param_2(parametrized_fixture):
    assert parametrized_fixture == param1 or parametrized_fixture == param2
    assert parametrized_fixture is param1 or parametrized_fixture is param2


def test_without_param_1(parametrized_fixture):
    assert parametrized_fixture == no_param
    assert parametrized_fixture is no_param


def test_without_param_2(parametrized_fixture):
    assert parametrized_fixture == no_param
    assert parametrized_fixture is no_param
```

When I run this, all tests pass. Note that I check both for equivalence *and* identity. So pytest keeps one instance per param (and one for no param).

Reply:
I don't think your test proves anything, it just shows how Python strings work. Take a look at this:
```
>>> a="hello"
>>> b=a
>>> b is a
True
>>> c=a
>>> c is a
True
```
As you can see, you can have several instances of a fixture but if they all "copy" (not really copy) the same string, "is" will tell you it's all the same string.

Anyway, my proposal was not to not use parametrize - I think you should continue using it. You would combine the best of both worlds. Something like this (incomplete and untested)

```
# The following are two session-scoped fixtures each only created once, ever
@pytest.fixture(scope="session")
def test_keyspace_with_tablets(cql, ...):
   # the code to create a table with tablets:
   ....

@pytest.fixture(scope="session")
def test_keyspace_with_vnodes(cql, ...):
   # the code to create a table with vnodes:
   ....

# This is a function-scope fixture, a separate one is "created" each time it's used,
# but there is no creation involved, we just return one of the two existing fixtures
@pytest.fixture(scope="function")
def test_keyspace(request, test_keyspace_with_tablets, test_keyspace_with_vnodes):
    if hasattr(request, "param"):
      if request.param == "vnodes":
            return test_keyspace_with_vnodes
        elif request.param == "tablets":
            return test_keyspace_with_tablets
        else:
            pytest.fail(f"test_keyspace(): invalid request parameter: {request.param}")
    cql.execute(stmt)
```

Reply:
Ah, I misunderstood your comment. Yes, your suggestion makes sense, I will incorporate it.

Reply:
Fixed in v2.